keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+   selector:
+     matchLabels:
+       app: wordpress
+       component: wordpress-webserver
+   template:
+     metadata:
+       labels:
+         app: wordpress
+         tier: frontend
+     spec:
+       containers:
+       - image: wordpress:4.8-apache
+         name: wordpress
+         env:
+         - name: WORDPRESS_DB_HOST
+           value: wordpress-mysql
+         - name: WORDPRESS_DB_PASSWORD
+           valueFrom:
+             secretKeyRef:
+               name: mysql-pass
+               key: password
+         ports:
+         - containerPort: 80
+           name: wordpress
+         volumeMounts:
+         - name: wordpress-persistent-storage
+           mountPath: /var/www/html
+   volumeClaimTemplates:
+        - metadata:
+            name: mysql-persistent-storage
+          spec:
+            accessModes: [ "ReadWriteOnce" ]
+            resources:
+              requests:
+                storage: 250Gi
+ ```
+ 
+Form an infrastructure management perspective, the manifest above encapsulates everything that is necessary to create 
+the workloads and services that compose our WordPress CMS application. However, the fact that these workloads comprise 
+an application is lost in translation during creation. Once these manifests are applied, there is no default way 
+for users to reconstruct the fact that the whole is more than the sum of its parts.
+ 
+When UIs, either command line or graphical, seek to interact with the application, they can only interact with its 
+components, and there is no standard way to indicate that those components comprise an application. As previously 
+mentioned, Many tools in the ecosystem attempt to aggregate the components of an application using labels and/or 
+annotations, but, as each tool uses a different scheme, an application created by one tool is generally not recognizable 
+by another, and UI designers have no standard way to recognize and display an application created by any tool.
+
+Moreover, their is no way to determine if the application is healthy. Even if a UI is able to determine the components 
+and dependencies of the application, and even if it is able to determine that the individual components are healthy 
+(e.g All Services are created and all Deployments are fully replicated), this does not necessarily imply that the 
+application as a whole is able to service requests.
+
+
+### Goals
+
+1. Provide a standard API for creating, viewing, and managing applications in Kubernetes.
+1. Provide a CLI implementation, via kubectl, that interacts with the Application API.
+1. Provide installation status and garbage collection for applications.
+1. Provide a standard way for applications to surface a basic health check to the UIs.
+1. Provide an explicit mechanism for applications to declare dependencies on another application.
+1. Promote interoperability among ecosystem tools and UIs by creating a standard that tools MAY implement.
+
+### Non-Goals
```

Comment:
What about package searching and browsing?

This contains some, but not all, of the properties needed for that use case, but it isn't mentioned as either a goal or non-goal.